lib/viewmodels/upvote_view_model.dart

```diff
@@ -0,0 +1,97 @@
+import "dart:async";
+
+import "package:hooks_riverpod/hooks_riverpod.dart";
+import "package:proxima/models/database/post/post_firestore.dart";
+import "package:proxima/models/database/post/post_id_firestore.dart";
+import "package:proxima/models/database/vote/upvote_state.dart";
+import "package:proxima/models/ui/post_vote.dart";
+import "package:proxima/services/database/post_repository_service.dart";
+import "package:proxima/services/database/post_upvote_repository_service.dart";
+import "package:proxima/viewmodels/login_view_model.dart";
+
+/// This view model is used to handle the upvoting and downvoting of a post.
+/// It exposes a [PostVote] object that contains the current vote state and the
+/// number of votes.
+/// The [postId] parameter is used to identify the post for which is responsible
+/// this view model.
+///
+/// To avoid unnecessary database fetches, the [PostVote] state is queried
+/// from the database only once at build time and then updated locally after the
+/// user triggers an upvote or downvote action.
+///
+/// This view model must be AutoDisposed because it is created on the fly
+/// for each post that is displayed on the screen.
+/// Thus, when the post is removed from the screen, the view model is disposed.
+class UpVoteViewModel extends AutoDisposeFamilyAsyncNotifier<PostVote,
+    ({PostIdFirestore postId})> {
```

Comment:
We should replace this type tuple by the actual `PostIdFirestore` type of its own. It does not really matter if it needs to be called `arg` on the overridden method.

Reply:
I had originally done that because I though it would made the code easier to understand.
But looking back at it, this adds quite some complexity and I believe that this doesn't really worth it actually. 
So thank you for pointing it out :+1:

Reply:
Yes, much better thx.